test/Main.hs

```diff
@@ -0,0 +1,208 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main where
+
+import Control.Exception (catch, SomeException)
+import Control.Monad (forM, unless)
+import qualified Data.ByteString.Lazy as LBS
+import Data.List (isPrefixOf, isSuffixOf, isInfixOf, sort, tails)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as TE
+import System.Directory
+  ( createDirectoryIfMissing
+  , doesFileExist
+  , getCurrentDirectory
+  , getTemporaryDirectory
+  , listDirectory
+  , removeFile
+  , setCurrentDirectory
+  )
+import System.Exit (ExitCode(..))
+import System.FilePath ((</>), dropExtension, makeRelative, takeBaseName, takeDirectory, takeFileName)
+import System.Process (readProcessWithExitCode)
+import Test.Tasty (TestTree, defaultMain, testGroup)
+import Test.Tasty.Golden (goldenVsStringDiff)
+
+-- | Convert a String to a lazy ByteString using UTF-8 encoding.
+stringToLBS :: String -> LBS.ByteString
+stringToLBS = LBS.fromStrict . TE.encodeUtf8 . T.pack
+
+main :: IO ()
+main = do
+  -- Discover the test directory path (relative to the current working directory)
+  cwd <- getCurrentDirectory
+  let testDir = cwd </> "test"
+
+  -- Change to the test directory so that agda2hs can find the agda-lib
+  setCurrentDirectory testDir
+
+  -- Create a temporary build directory
+  tmpDir <- getTemporaryDirectory
+  let buildDir = tmpDir </> "agda2hs-test-build"
+  createDirectoryIfMissing True buildDir
+
+  -- Discover test cases
+  succeedTests <- discoverSucceedTests testDir buildDir
+  failTests <- discoverFailTests testDir buildDir
+
+  -- Run tests
+  defaultMain $ testGroup "agda2hs"
+    [ testGroup "Succeed" succeedTests
+    , testGroup "Fail" failTests
+    ]
+
+-- | Discover all .agda files under the Succeed directory recursively.
+discoverSucceedTests :: FilePath -> FilePath -> IO [TestTree]
+discoverSucceedTests testDir buildDir = do
+  agdaFiles <- findAgdaFilesRecursive (testDir </> "Succeed")
+  forM (sort agdaFiles) $ \agdaFile -> do
+    let testName = dropExtension (makeRelative (testDir </> "Succeed") agdaFile)
+        goldenFile = dropExtension agdaFile ++ ".hs"
+    return $ succeedTest testDir buildDir testName agdaFile goldenFile
+
+-- | Discover all .agda files under the Fail directory.
+discoverFailTests :: FilePath -> FilePath -> IO [TestTree]
+discoverFailTests testDir buildDir = do
+  agdaFiles <- findAgdaFilesRecursive (testDir </> "Fail")
+  forM (sort agdaFiles) $ \agdaFile -> do
```

Comment:
Again: The files should be ordered first by modification date (newest first) and then by modification date of their golden value (if it exists), then alphabetically.